<div style="background-color: #add8e6; padding: 10px; height: 70px; border-radius: 15px;">
    <div style="font-family: 'Georgia', serif; font-size: 20px; padding: 10px; text-align: right; position: absolute; right: 20px;">
        Mohammad Idrees Bhat <br>
        <span style="font-family: 'Arial', sans-serif;font-size: 12px; color: #0a0a0a;">Tech Skills Trainer | AI/ML Consultant</span> <!--- Mohammad Idrees Bhat | Tech Skills Trainer | AI/ML Consultant --->
    </div>
</div>

<!--- Mohammad Idrees Bhat | Tech Skills Trainer | AI/ML Consultant --->

<div style="background-color: #002147; padding: 10px; text-align: center; color: white; font-size: 32px; font-family: 'Arial', sans-serif;">
     Model Refinement <br>
    <h3 style="text-align: center; color: white; font-size: 15px; font-family: 'Arial', sans-serif;">Improving model performance based on evaluation results</h3>
</div>

<div style="background-color: white; color: black; padding: 10px;">
    <h4><b>AGENDA</b> <p><p>
1.  Understanding Model Training Errors<p><p> 
2.  Techniques for Model Refinement <p>
3.  Advanced Refinement Techniques <p>
4.  Hands-On: Model Refinement Practice  <p>  
</h4> </div>

<!-- Link the Montserrat font -->
<link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@700&display=swap" rel="stylesheet">

<!-- Main div with centered content and a flexible box size, no scroll bar -->
<div style="background-color: #baf733; min-height: 100px; width: 100%; display: flex; justify-content: center; align-items: center; position: relative; padding: 20px; box-sizing: border-box; font-family: 'Montserrat', sans-serif; font-weight: 700; font-size: 20px; border-radius: 15px;">
    <div style="position: absolute; top: 10px; right: 10px; padding: 5px 10px; font-size: 14px; color: rgba(0, 0, 0, 0.05); border-radius: 10px;">Mohammad Idrees Bhat</div>
    <!-- Fill the below text with question -->
    <!-- Fill the below text with question -->
    If you were a superhero, what would your superpower be?
    <!-- Fill the above text with question -->
    <!-- Fill the above text with question -->

</div>

By the end of this session, students will be able to:

- Evaluate key performance metrics for model assessment.
- Identify and address issues of underfitting and overfitting in models.
- Apply a structured process of data preprocessing, hyperparameter tuning, and ensemble methods for model improvement.
- Confidently implement refinement techniques to boost model accuracy and robustness.


**Evaluation Metrics Recap:**

**Core Metrics**
- Accuracy: The proportion of correct predictions out of the total predictions. Often used when class distributions are balanced.
- Precision: The ratio of correctly predicted positive observations to the total predicted positives. Precision is critical when false positives are costly (e.g., spam detection).
- Recall: The ratio of correctly predicted positive observations to all observations in the actual class. Useful in cases where capturing all positives is crucial, like detecting diseases.
- F1-Score: The harmonic mean of precision and recall. Balances both metrics, making it ideal when both false positives and false negatives are important.

**Choosing the Right Metric**
It is important to select the right metric that ensures the model aligns with the real-world requirements of a given problem:

- **Classification**: Use AUC in credit fraud detection to balance detection rates with false positives.
- **Regression**: Use RMSE in forecasting where larger errors need penalization, such as in stock price predictions.

<div style="background-color: lightgreen; color: black; padding: 4px;">
    <h3>1. Understanding Model Training Errors
</h3> </div>

In machine learning, errors are an inevitable part of building and refining models.

Understanding where these errors come from and how to manage them is essential for creating effective models that perform well on new, unseen data. 

When we train a machine learning model, we want it to be as accurate as possible—not just on the training data, but also on the data it has never seen before. This is where the generalization of a model comes into play. 

Good generalization means the model can perform well on unseen data, and is not just memorizing the training examples.

Some of the most important model training errors are:

1. **Bias:**



Imagine you're trying to guess the height of a group of people based on their age. 

If you always guess the same height for everyone, no matter what their age is, that would be a biased guess. 

You're not considering their actual ages and just making a simple guess based on some wrong assumption.


*Simple Example*

Let's say you’re using a machine learning model to predict house prices based on their size (in square feet). If your model always predicts the same price for every house, regardless of its size, that would be high bias. The model is too simple and doesn't take enough details into account. It’s not learning the true relationship between house size and price.



- Bias refers to systematic errors in the model, introduced by approximating a real-world problem, which may be complex, by a simplified model.

- It occurs when the model makes assumptions that are too simple, often ignoring important features or relationships in the data.

- High bias typically results in underfitting, where the model is too simple to capture the underlying patterns of the data.

- Example: Using a linear regression model to predict house prices when the true relationship is non-linear (e.g., polynomial).

![](https://static.javatpoint.com/tutorial/machine-learning/images/bias-and-variance-in-machine-learning6.png)

2. **Variance:**



Variance is like a model's sensitivity to small changes in the data. 

Imagine you're trying to guess the height of people based on their age. If you adjust your guess every time you see a new person, even when they’re about the same age, then your guesses might be all over the place — that’s high variance!

In machine learning, high variance means the model is trying too hard to fit every little detail in the training data, even the noise or random patterns that aren’t important.

*Simple Example*
Suppose you have a model that predicts house prices based on the square footage and several other features. If your model is high in variance, it might make big adjustments for every single small change in these features. So, a house with just a tiny difference in square footage might get a drastically different price prediction, even if that change shouldn't really matter much.

- Variance refers to how much the model's predictions fluctuate when trained on different subsets of data.
- It occurs when the model is too complex and overly sensitive to the noise or fluctuations in the training data.
- High variance typically leads to overfitting, where the model fits the training data too well but fails to generalize to new, unseen data.
- Example: A decision tree with many branches that memorizes the training data instead of learning the general patterns.

So, you want to find a balance where the model is not too sensitive (high variance) but also not too simple (high bias). This helps the model perform well on both the training data and any new data it sees.

**Bias-Variance Tradeoff**

- The Bias-Variance Tradeoff refers to the tension between these two sources of error. 

As you reduce bias (by using a more complex model), variance tends to increase (because the model is more sensitive to fluctuations in the training data), and vice versa.

- Goal: Find the optimal balance between bias and variance, where the model has both low bias and low variance, resulting in good generalization to new data.

![](https://miro.medium.com/max/1400/1*wPUGn4buYw4LYISGL-TUuA.png)

**3. Underfitting vs. Overfitting**

Underfitting occurs when the model is too simple and cannot capture the underlying patterns in the data. This leads to poor performance on both the training set and the test set.
- High Bias, Low Variance: Simple models (underfit).

Overfitting occurs when the model is too complex and captures not just the true patterns but also the noise in the data. This leads to excellent performance on the training set but poor generalization to new, unseen data
- Low Bias, High Variance: Complex models (overfit).

**4. Data Imbalance**

Data imbalance happens when you have unequal numbers of examples in different classes of your dataset. 

For example, if you’re building a model to detect fraud, you might have a dataset where 95% of transactions are normal and only 5% are fraudulent. This difference is called imbalanced data.

- Bias Toward Majority Class: When data is imbalanced, the model may become biased towards the majority class (the larger group). So, in the fraud example, the model might mostly predict "normal" because it’s the majority. This can cause the model to ignore or perform poorly on the minority class (fraudulent transactions).

- Poor Accuracy on on Minority Class: If the minority class (fraud) is more important to predict, imbalance can be especially problematic. Even if the model achieves high overall accuracy by predicting the majority class, it might fail to detect the minority class, which is often the class we care most about.

In real life, we often fix data imbalance by resampling the data 

— either by **oversampling** the minority class or **undersampling** the majority class 

— and instead of just accuracy, use metrics like Precision, Recall, and the F1 score, which give a better picture of how well the model is performing on each class. 

Additionally, special algorithms like SMOTE (Synthetic Minority Over-sampling Technique) can create synthetic samples to help balance the dataset.



**5. Data Drift**

Data Drift happens when the data used to train a model changes over time, so it no longer matches the new data the model sees in the real world. This can lead to poor performance because the model is making predictions based on outdated patterns.

**Reasons:**

- Changing User Behavior: For example, in a recommendation system, user preferences might shift over time.
- Seasonal Changes: Data may change with the seasons or economic conditions (like shopping trends during holidays).
- System Updates: Changes in how data is collected or processed (like updates to sensors or apps) can also cause drift.

**Types of Data Drift:**

1. **Covariate Drift:** When the input data (features) distribution changes but the relationship to the target variable remains the same.

**Example:** 
Let’s say a model was trained to assess creditworthiness based on features like age, income, and spending habits. 

Over time, if economic conditions change (like during a recession), people's spending habits and income levels may shift, even though the relationship between these features and credit risk (the target variable) remains the same.

2. **Concept Drift:** When the relationship between inputs and the target variable itself changes. 

**Example:**
Suppose a model is trained to classify emails as spam based on specific keywords and patterns. Over time, as spammers develop new tactics, the patterns in spam emails change (e.g., new words, different structures).

Now, the original relationship between certain keywords and the label "spam" no longer holds, meaning the model becomes less effective. This change in the underlying relationship between inputs (keywords) and the target (spam/not spam) is concept drift.

In real life, we handle data drift by regularly monitoring model performance, retraining the model with updated data, and using adaptive models that adjust to new data patterns.

<div style="background-color: lightblue; color: black; padding: 4px;">
    <h4> Model Refinement
</h4> </div>

Model Refinement in machine learning refers to the process of improving a model's performance by iterating on it through various techniques. 

This step is crucial to ensure that the model generalizes well to new, unseen data

The model refinement process typically includes:

1. **Evaluating Model Performance:** Carefully analyzing model metrics to identify where the model is falling short.
2. **Identifying Key Issues:** Pinpointing problems like high bias (indicating underfitting), high variance (overfitting), or data imbalance.
3. **Applying Improvement Techniques:** Using targeted methods, such as tuning hyperparameters, changing model architecture, or refining the dataset, to tackle identified issues.
4. **Iterative Testing and Validation:** Running and evaluating improvements in multiple rounds to ensure they lead to better and more stable results.

<div style="background-color: lightgreen; color: black; padding: 4px;">
    <h3>2. Techniques for Model Refinement
    </h3> </div>

Some Techniques to systematically refine your models and make their outputs and predictions better. 

<div style="background-color: lightblue; color: black; padding: 4px;">
    <h4> Step 1: Data Preprocessing Refinement
</h4> </div>

In this first step of model refinement, we’ll focus on preparing the data to improve model performance. Well-preprocessed data ensures that models learn effectively, reducing bias and variance in the results.

### **Data Scaling and Normalization:**

In scaling (Standardization), the goal is to standardize features so they have a mean of 0 and standard deviation of 1


- **When to Scale:** Scaling techniques like StandardScaler or MinMaxScaler are critical when using algorithms sensitive to feature magnitudes, such as distance-based algorithms (e.g., K-Nearest Neighbors) and linear models.
- **How it Helps:** Scaling or normalizing features to the same range or distribution can improve model convergence, stability, and predictive performance.

In Normalization, the goal is to transforms features so they fall within a certain range, typically [0, 1] or [-1, 1].

More or less the same, but normalization **adjusts the range of data** to a specific interval (e.g., [0, 1]), while **scaling** **adjusts the spread** or **distribution** of the data, typically by changing the mean and standard deviation.

**Example**

Suppose we have a dataset with the following features to predict house prices:

- **Size** (in square feet)
- **Number of Rooms**
- **Distance to City Center** (in miles)

Here’s a sample of what the dataset might look like:

| Size (sq ft) | Number of Rooms | Distance to City Center (miles) | Price ($) |
|--------------|-----------------|----------------------------------|-----------|
| 2500         | 4               | 8                                | 500,000   |
| 1500         | 3               | 15                               | 300,000   |
| 1200         | 2               | 20                               | 200,000   |
| 1800         | 3               | 10                               | 350,000   |

Without scaling, the large range of values (e.g., Size in thousands vs. Distance in tens) may lead some models to interpret Size as more important than other features. Here’s how scaling can address this:

**Standard Scaling**

Standard scaling (using `StandardScaler`) adjusts each feature so that it has a mean of 0 and a standard deviation of 1.

**Transformed dataset (Standard Scaled):**

| Size (scaled) | Number of Rooms (scaled) | Distance to City Center (scaled) |
|---------------|--------------------------|-----------------------------------|
| 1.14          | 1.0                      | -0.6                             |
| -0.2          | 0.0                      | 0.9                              |
| -0.8          | -1.0                     | 1.5                              |
| -0.1          | 0.0                      | -0.3                             |

**Min-Max Scaling**

Min-Max scaling (using `MinMaxScaler`) transforms each feature to a specified range, typically [0, 1].

**Transformed dataset (Min-Max Scaled):**

| Size (scaled) | Number of Rooms (scaled) | Distance to City Center (scaled) |
|---------------|--------------------------|-----------------------------------|
| 1.0           | 1.0                      | 0.0                              |
| 0.3           | 0.5                      | 0.54                             |
| 0.0           | 0.0                      | 1.0                              |
| 0.5           | 0.5                      | 0.31                             |

**Why This Matters**

- **Improved Model Performance**: Models sensitive to feature scales, like K-Nearest Neighbors and Linear Regression, perform better with scaled features.
- **Faster Convergence**: Gradient-based models, like logistic regression or neural networks, converge faster when features are on a similar scale.

*Find code implementation in the "DataScaling and Normalisation.ipynb"*

**Techniques Used in Scaling**

- Standard Scaling (Z-score Normalization)
- Robust Scaling
- Max Abs Scaling

**Techniques Used in Normalization**

- Min-Max Normalization (Rescaling)
- L2 Normalization (Euclidean norm)
- L1 Normalization (Manhattan norm)


---

### **Feature Engineering:**

- **Feature Selection:** Removing features with high correlation or low variance can reduce redundancy and prevent the model from overemphasizing irrelevant or repetitive information.
- **Feature Creation/Extraction:** New features can be engineered by transforming or combining existing ones, potentially uncovering valuable patterns that the model might otherwise miss.

**Handling Class Imbalance:**
- **Why it Matters:** In classification problems, class imbalance can cause models to perform poorly on the minority class, as they may simply predict the majority class to achieve high accuracy.
- **Techniques to Address Imbalance:**
    - **Oversampling** (e.g., duplicating instances of the minority class).
    - **Undersampling** (e.g., reducing the majority class).
    - **SMOTE (Synthetic Minority Over-sampling Technique)**, which creates synthetic samples for the minority class.

<div style="background-color: lightblue; color: black; padding: 4px;">
    <h4> Hyperparameter Tuning
</h4> </div>

Hyperparameters are configurations **external to the model** that cannot be learned from data during training. They differ from parameters, which are learned by the model.

For instance, in a **decision tree**, **parameters** include **split points** within nodes, whereas **hyperparameters** may include the **tree depth** or **minimum number of samples** required to split a node.

Proper tuning of hyperparameters can significantly enhance the model's accuracy, generalizability, and training efficiency.

**Hyperparameters vs. Parameters**

- **Parameters**: Learned directly from the data (e.g., weights in a linear regression model).
- **Hyperparameters**: Set manually before training begins and control the behavior of the learning algorithm (e.g., k in k-nearest neighbors).

---

**Hyperparametes for some commonly used algorithms**

1. **Linear Regression**
   - **Hyperparameters**: 
     - Regularization strength (for Ridge and Lasso: alpha)
     - Solver type (e.g., ‘lbfgs’, ‘saga’)

2. **Logistic Regression**
   - **Hyperparameters**: 
     - Regularization strength (C)
     - Solver type (e.g., ‘liblinear’, ‘newton-cg’)
     - Max iterations

3. **Decision Trees**
   - **Hyperparameters**:
     - Maximum depth (max_depth)
     - Minimum samples split (min_samples_split)
     - Minimum samples leaf (min_samples_leaf)
     - Max features

4. **Random Forest**
   - **Hyperparameters**:
     - Number of trees (n_estimators)
     - Max depth (max_depth)
     - Max features
     - Minimum samples per leaf (min_samples_leaf)
     - Bootstrap (for sampling)

5. **Support Vector Machines (SVM)**
   - **Hyperparameters**:
     - Regularization parameter (C)
     - Kernel type (e.g., ‘linear’, ‘rbf’, ‘poly’)
     - Gamma (for RBF kernel)
     - Degree (for polynomial kernel)

6. **K-Nearest Neighbors (KNN)**
   - **Hyperparameters**:
     - Number of neighbors (n_neighbors)
     - Distance metric (e.g., ‘euclidean’, ‘manhattan’)
     - Weights (uniform or distance)

7. **Gradient Boosting Machines (GBM) / XGBoost / LightGBM**
   - **Hyperparameters**:
     - Number of trees (n_estimators)
     - Learning rate (learning_rate)
     - Maximum depth (max_depth)
     - Subsample
     - Min child weight (XGBoost)

8. **Naive Bayes**
   - **Hyperparameters**:
     - Smoothing parameter (alpha)

9. **K-Means Clustering**
   - **Hyperparameters**:
     - Number of clusters (n_clusters)
     - Initialization method (init: ‘k-means++’ or ‘random’)
     - Max iterations (max_iter)
     - Tolerance for convergence (tol)

10. **Neural Networks (e.g., MLPClassifier)**
    - **Hyperparameters**:
      - Number of hidden layers (hidden_layer_sizes)
      - Number of neurons per layer
      - Activation function (e.g., ‘relu’, ‘tanh’)
      - Learning rate (learning_rate)
      - Batch size
      - Epochs (max_iter)
      - Solver type (e.g., ‘adam’, ‘sgd’)

11. **Principal Component Analysis (PCA)**
    - **Hyperparameters**:
      - Number of components (n_components)
      - Whitening (True/False)

---

**Common Hyperparameter Tuning Techniques**

**1. Grid Search:**

This method tests all possible combinations of a predefined set of hyperparameter values exhaustively. Though powerful, it can be computationally expensive.

Example: Testing combinations of max_depth and min_samples_split for a decision tree.

**2. Random Search:**

Instead of testing all combinations, random search samples a fixed number of hyperparameter combinations randomly from a specified distribution. This often finds optimal settings more efficiently than grid search.

Example: Randomly selecting learning_rate and n_estimators values for a gradient boosting model.

<div style="background-color: lightblue; color: black; padding: 4px;">
    <h4> Regularization Techniques
</h4> </div>

What is Regularization? (Brief overview)
Types of Regularization:
L1 Regularization (Lasso): Tends to produce sparse models by penalizing high coefficients.
L2 Regularization (Ridge): Helps reduce overfitting by penalizing large coefficients but keeps all features.
Activity: Students add L1 or L2 regularization to a model in code and compare results before and after regularization.

<div style="background-color: lightgreen; color: black; padding: 4px;">
    <h3>3. Advanced Refinement Techniques
    </h3> </div>

<div style="background-color: lightblue; color: black; padding: 4px;">
    <h4> Ensemble Methods 
</h4> </div>

Step 3: Ensemble Methods
Ensemble Learning Introduction:
Explain how ensembles (e.g., Bagging, Boosting, Stacking) can improve model performance by combining multiple models.
Examples:
Bagging: Random Forest
Boosting: AdaBoost, Gradient Boosting, XGBoost
Stacking: Combining different models (e.g., Decision Tree, SVM) into a final meta-model.
Practical Exercise: Implement a simple ensemble model and observe performance gains over individual models.

Step 4: Cross-Validation and Model Validation Techniques
K-Fold Cross-Validation:
Recap the importance of cross-validation in ensuring that model performance is not overestimated.
Practical Exercise: Have students apply K-Fold cross-validation and calculate average performance metrics.
Advanced Cross-Validation (e.g., Stratified K-Fold, Time Series Split):
Discuss when and how to use these techniques for imbalanced or time-based data.
Train-Test-Split Best Practices:
Ensure students understand the need for a proper train-test split and common pitfalls (e.g., leakage between training and test sets).

<div class="alert alert-block alert-warning">
    <b><font size="5"> 4. Hands-On Mini-Project: Model Refinement (Optional) </font> </b>
</div>

<!--- Mohammad Idrees Bhat | Tech Skills Trainer | AI/ML Consultant --->

### **Project Title: Customer Churn Prediction Model Refinement Challenge**

**Description**

This project centers on **predicting customer churn**, a popular real-world problem in industries like **telecom**, **SaaS**, or **e-commerce**. The dataset contains information about customer demographics, subscription details, service usage, and customer support interactions. You will build a machine learning model to predict whether a customer will churn (i.e., stop using the service) based on these features.

This project is modern, widely applicable, and gives you insight into solving a business-critical problem.

**Instructions**

1. **Define the Problem**: Predict customer churn using the Telco Customer Churn dataset.  
2. **Baseline Model**: Train a simple classifier with minimal preprocessing.  
3. **Evaluate Baseline**: Analyze accuracy, recall, and other key metrics.  
4. **Data Preprocessing**: Handle missing values, encode categorical features, and scale numerical data.  
5. **Feature Engineering**: Create new features, remove redundant ones, and analyze correlations.  
6. **Hyperparameter Tuning**: Optimize model parameters using GridSearchCV or similar tools.  
7. **Model Ensembling**: Use ensemble techniques like Gradient Boosting or Random Forest to enhance performance.  
8. **Final Evaluation**: Compare refined model metrics with baseline results and assess business value.  
9. **Presentation**: Summarize findings, improvements, and rationale for techniques used.  

Now it's your turn!
### Task 1: description of task

    - instructions

<div style="background-color: lightblue; color: white; padding: 10px; text-align: center;">
    <h1>_________________________________END________________________________
        <!--- Mohammad Idrees Bhat | Tech Skills Trainer | AI/ML Consultant --->
</h1> </div>

<div style="background-color: #002147; color: #fff; padding: 30px; text-align: center;">
    <h1>THANK YOU!
        <!--- Mohammad Idrees Bhat | Tech Skills Trainer | AI/ML Consultant --->
</h1> </div>

<div style="background-color: lightgreen; color: black; padding: 30px;">
    <h4> Live Exercise Solutions
        
</h4> </div>

**Step 1: Define the Problem**

**Objective**: Predict whether a customer will churn (`Churn = 1`) or not (`Churn = 0`) using the Telco Customer Churn dataset.

- Dataset: Telco Customer Churn (`Telco-Customer-Churn.csv`).
- Target Variable: `Churn`.
- Features: Various customer attributes such as contract type, payment method, and internet service.

In [1]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

Use the line below to download and import the dataset

In [2]:
# pip install kagglehub # if kagglehub is not installed yet
#import kagglehub

# Download latest version
#path = kagglehub.dataset_download("blastchar/telco-customer-churn")

#print("Path to dataset files:", path)

- A link to the dataset : [Telco Customer Churn](https://www.kaggle.com/datasets/blastchar/telco-customer-churn)

In [3]:
# Step 2: Load dataset

# Load the dataset
# Use the path to get to the dataset
# Rename the dataset and choose the right path
import pandas as pd

data = pd.read_csv('c:\\Users\\devid\\.cache\\kagglehub\\datasets\\blastchar\\telco-customer-churn\\versions\\1\\Telco-Customer-Churn.csv')


In [4]:
# Step 3: Data Preprocessing

# Minimal preprocessing: Handle missing values and encode target variable
data = data.dropna()  # Drop missing values
data['Churn'] = data['Churn'].apply(lambda x: 1 if x == 'Yes' else 0)  # Encode target variable

# Replace empty strings or whitespace in string columns with NaN and drop them
data.replace(r'^\s*$', pd.NA, regex=True, inplace=True)  # Replace empty strings with NaN
# data.dropna(inplace=True)  # Drop any rows with NaN values

In [5]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OrdinalEncoder, OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder

# Define column categories
categorical_cols = ['Contract', 'PaymentMethod', 'InternetService', 'MultipleLines']
numerical_cols = ['tenure', 'MonthlyCharges', 'TotalCharges']
binary_cols = ['gender', 'Partner', 'Dependents', 'PhoneService', 
               'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 
               'TechSupport', 'StreamingTV', 'StreamingMovies', 'PaperlessBilling']

# Define preprocessing for binary columns
binary_preprocessor = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),  # Handle missing values
    ('ordinal', OrdinalEncoder())  # Ordinal encode binary columns
])

# Define preprocessing for other categorical columns
categorical_preprocessor = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),  # Handle missing values
    ('onehot', OneHotEncoder(handle_unknown='ignore'))     # One-hot encode
])

# Define preprocessing for numerical columns
numerical_preprocessor = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),           # Handle missing values
    ('scaler', StandardScaler())                           # Scale numerical features
])

# Combine into a single preprocessor
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_preprocessor, numerical_cols),
        ('cat', categorical_preprocessor, categorical_cols),
        ('binary', binary_preprocessor, binary_cols)  # Add binary columns processing
    ],
    remainder='passthrough'  # Leave any remaining columns unchanged
)

In [6]:
# Apply to the dataset
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression

# Create the model pipeline
model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(max_iter=1000))
])

In [8]:
# Fit the pipeline to your data
# Note: Ensure 'customerID' is removed before fitting
X = data.drop(['Churn', 'customerID'], axis=1)
y = data['Churn']



In [9]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

model.fit(X_train, y_train)

TypeError: float() argument must be a string or a real number, not 'NAType'

In [10]:
print(X_train.isnull().sum())  # Check for missing values in the features
print(y_train.isnull().sum())  # Check for missing values in the target


gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        8
dtype: int64
0


In [11]:
# Convert 'TotalCharges' to numeric, coercing errors into NaN
data['TotalCharges'] = pd.to_numeric(data['TotalCharges'], errors='coerce')

# Now fill missing values with the mean
data['TotalCharges'].fillna(data['TotalCharges'].mean(), inplace=True)

# Double-check to ensure there are no missing values left in 'TotalCharges'
print(data['TotalCharges'].isnull().sum())  # Should print 0 after filling

0


C:\Users\devid\AppData\Local\Temp\ipykernel_25300\2174509298.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['TotalCharges'].fillna(data['TotalCharges'].mean(), inplace=True)


In [12]:
# Fit the pipeline to your data
# Note: Ensure 'customerID' is removed before fitting
X = data.drop(['Churn', 'customerID'], axis=1)

In [13]:
# Train-test split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Train the model
model.fit(X_train, y_train)

# Evaluate the model
from sklearn.metrics import classification_report
y_pred = model.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.90      0.88      1539
           1       0.69      0.57      0.62       574

    accuracy                           0.81      2113
   macro avg       0.77      0.74      0.75      2113
weighted avg       0.81      0.81      0.81      2113



In [39]:
y_train.value_counts()

Churn
0    3635
1    1295
Name: count, dtype: int64

### **Let's now try to improve these results**

5. Feature Engineering:
Interaction Features: Create new features that might capture relationships between existing features. For example, interactions between tenure and MonthlyCharges might reveal insights about customer retention.
Polynomial Features: You can try adding polynomial features for numerical variables. This can help capture non-linear relationships.

In [14]:
from sklearn.preprocessing import PolynomialFeatures

# Add polynomial features for numerical columns
# Create a pipeline to process numerical columns by adding polynomial features
polynomial_preprocessor = Pipeline(steps=[

    # Step 1: Imputer
    # This step handles missing values in the numerical data by replacing them with the mean value of the column.
    # It ensures that the polynomial feature generation doesn't fail due to missing values.
    ('imputer', SimpleImputer(strategy='mean')),

    # Step 2: PolynomialFeatures
    # This step generates polynomial features (up to degree 2 in this case) from the numerical data.
    # It allows the model to capture non-linear relationships by creating interaction terms and squared terms for each feature.
    ('poly', PolynomialFeatures(degree=2, include_bias=False)),

    # Step 3: StandardScaler
    # This step standardizes the features by scaling them to have a mean of 0 and standard deviation of 1.
    # This is important for many machine learning algorithms, as it ensures that all features are on the same scale.
    ('scaler', StandardScaler())
])

**6. Hyperparameter Tuning:**

Grid Search or Randomized Search: This is a very important step in improving model performance. Grid search helps identify the best hyperparameters for the Logistic Regression model.

In [15]:
# create a parameter grid for logistic regression
param_grid = {
    'classifier__C': [0.1, 1, 10],  # Regularization strength
    'classifier__solver': ['liblinear', 'saga'],  # Solvers to try
    'classifier__penalty': ['l1', 'l2'],  # Regularization type
}

from sklearn.model_selection import GridSearchCV

grid_search = GridSearchCV(model, param_grid, cv=5, scoring='accuracy', n_jobs=-1)
grid_search.fit(X_train, y_train)

# Get the best parameters and fit the model
print("Best parameters found: ", grid_search.best_params_)
model = grid_search.best_estimator_


Best parameters found:  {'classifier__C': 10, 'classifier__penalty': 'l1', 'classifier__solver': 'saga'}


In [16]:
# Directly use the best estimator from GridSearchCV
best_model = grid_search.best_estimator_

# Fit the best model (if not already fitted during grid search)
best_model.fit(X_train, y_train)

# Evaluate the model
y_pred = best_model.predict(X_test)

# Display the classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.85      0.90      0.88      1539
           1       0.69      0.57      0.62       574

    accuracy                           0.81      2113
   macro avg       0.77      0.74      0.75      2113
weighted avg       0.81      0.81      0.81      2113



The classification report gives a summary of key metrics for evaluating classification models, including precision, recall, f1-score, and support for each class in the dataset.

| Class   | Precision | Recall | F1-Score | Support |
|---------|-----------|--------|----------|---------|
| **0**   | 0.85      | 0.90   | 0.88     | 1539    |
| **1**   | 0.69      | 0.57   | 0.62     | 574     |
| **Accuracy**   |       |        | 0.81     | 2113    |
| **Macro avg**  | 0.77      | 0.74   | 0.75     | 2113    |
| **Weighted avg** | 0.81      | 0.81   | 0.81     | 2113    |


- For class 0, precision is 0.85, meaning 85% of predicted 0s were actually 0s.
- For class 1, precision is 0.69, meaning 69% of predicted 1s were actually 1s.

- Support refers to the number of occurrences of each class in the dataset.
 - Class 0 has 1539 instances, while class 1 has 574 instances. Which suggests class imbalance.

- Macro average computes the metrics for each class and then averages them. It does not take class imbalance into account.



Let's improve class balance and observe what happens.

By setting `class_weight='balanced'`, you're telling the model to adjust the weights inversely proportional to class frequencies. 

This means that the model will treat misclassifications of the minority class (class 1) as more costly and attempt to improve its performance on that class. 

However, this does not alter the support values in the output, which are based on the actual distribution of the classes in your dataset.

In [17]:
best_model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(max_iter=1000, class_weight='balanced'))
])

param_grid = {
    'classifier__C': [0.1, 1, 10],  # Regularization strength
    'classifier__solver': ['liblinear', 'saga'],  # Solvers to try
    'classifier__penalty': ['l1', 'l2'],  # Regularization type
}

# Perform Grid Search with cross-validation
grid_search = GridSearchCV(best_model, param_grid, cv=5, scoring='accuracy', n_jobs=-1)
# Fit the grid search model
grid_search.fit(X_train, y_train)

# Get the best parameters and fit the model
print("Best parameters found: ", grid_search.best_params_)
best_model2 = grid_search.best_estimator_

Best parameters found:  {'classifier__C': 0.1, 'classifier__penalty': 'l2', 'classifier__solver': 'liblinear'}


In [18]:
# Directly use the best estimator from GridSearchCV
best_model2 = grid_search.best_estimator_

# Fit the best model (if not already fitted during grid search)
best_model2.fit(X_train, y_train)

# Evaluate the model
y_pred = best_model2.predict(X_test)

# Display the classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.92      0.73      0.81      1539
           1       0.54      0.84      0.65       574

    accuracy                           0.76      2113
   macro avg       0.73      0.78      0.73      2113
weighted avg       0.82      0.76      0.77      2113



Let's Handle Class Imbalance More Effectively with 
SMOTE (Synthetic Minority Over-sampling Technique):

SMOTE requires numeric values for resampling. So let's fix our values in `y`



In [30]:
from imblearn.over_sampling import SMOTE
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.metrics import classification_report

# Apply preprocessing to training data
X_train_transformed = preprocessor.fit_transform(X_train)

# Step 1: Apply SMOTE to balance the classes in the training data
smote = SMOTE(random_state=42)
X_train_smote, y_train_smote = smote.fit_resample(X_train_transformed, y_train)

In [33]:
# Class distribution in y_train
import pandas as pd

# If y_train_smote is a NumPy array, convert it to a pandas Series for easier inspection
y_train_series = pd.Series(y_train)

# Get class distribution
class_distribution = y_train_series.value_counts()

print("Class distribution in y_train:")
print(class_distribution)

Class distribution in y_train:
Churn
0    3635
1    1295
Name: count, dtype: int64


In [34]:
# Class distribution in y_train_smote
import pandas as pd

# If y_train_smote is a NumPy array, convert it to a pandas Series for easier inspection
y_train_smote_series = pd.Series(y_train_smote)

# Get class distribution
class_distribution = y_train_smote_series.value_counts()

print("Class distribution in y_train_smote:")
print(class_distribution)

Class distribution in y_train_smote:
Churn
0    3635
1    3635
Name: count, dtype: int64


In [40]:
# Step 2: Define the model with balanced class weights
model = LogisticRegression(max_iter=1000, class_weight='balanced')

# Step 3: Train the model on the SMOTE-balanced data
model.fit(X_train_smote, y_train_smote)

# Step 4: Predict on the transformed test set
X_test_transformed = preprocessor.transform(X_test)
y_pred_smote = model.predict(X_test_transformed)

# Step 5: Evaluate the model and print the classification report to check support values
print(classification_report(y_test, y_pred_smote))

              precision    recall  f1-score   support

           0       0.92      0.73      0.81      1539
           1       0.53      0.83      0.65       574

    accuracy                           0.76      2113
   macro avg       0.73      0.78      0.73      2113
weighted avg       0.82      0.76      0.77      2113



**7. Ensemble Methods:**

Random Forest: A Random Forest classifier or another ensemble model might work better since it can capture more complex patterns.
Gradient Boosting: A boosting method like GradientBoostingClassifier or XGBoost often provides better results by focusing on the mistakes made by previous models.

Example: Replacing Logistic Regression with RandomForestClassifier:

In [ ]:
from sklearn.ensemble import RandomForestClassifier

model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', RandomForestClassifier(n_estimators=100, random_state=42))
])


 use Gradient Boosting:

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42))
])

<div class="alert alert-block alert-warning"  padding: 10px; text-align: center;">
    <font size="3"> Programming Interveiw Questions</font>
</div>

1. topic:
    - question

<!--- Mohammad Idrees Bhat | Mohammad Idrees Bhat --->

<h2 style="background-color: #ffe4e1; color: #2f4f4f; padding: 10px; border-radius: 10px; width: 350px; text-align: center; float: right; margin: 20px 0;">
    Mohammad Idrees Bhat<br>
    <span style="font-size: 12px; color: #696969;">
        Tech Skills Trainer | AI/ML Consultant
    </span>
</h2>

<!--- Mohammad Idrees Bhat | Tech Skills Trainer | AI/ML Consultant --->